File: 20_number_string/number_string.py
'''
  面试题20：表示数值的字符串
  题目：请实现一个函数用来判断字符串是否表示数值（包括整数和小数）。例如，
  字符串“+100”、“5e2”、“-123”、“3.1416”及“-1E-16”都表示数值，但“12e”、
  “1a3.14”、“1.2.3”、“+-5”及“12e+5.4”都不是
'''
def is_numeric(string):
  '''
  有限状态机
  '+-', 只能出现在字符串首元素，或者E/e后面
  '.', 只能出现一次，且不能出现在E/e后面
  'Ee', 只能出现一次，且其前后都需要有数字部分
  'digit', 数字部分
  '''
  # 判断输入有效
  if not string:
    return False
  
  string = string.strip()
  met_dot = met_e = met_digit = False

  for idx, char in enumerate(string):
    if char in ['+', '-']:
      if idx > 0 and string[idx-1] not in 'Ee':
        return False
    elif char == '.':
      if met_dot or met_e:
        return False
      met_dot = True
    elif char in ['E', 'e']:
      if met_e or not met_digit:
        return False
      met_e, met_digit = True, False
    elif char.isdigit():
      met_digit = True
    else:
      return False
  
  return met_digit

def Test(test, string, expected):
  print(test, is_numeric(string), expected)

if __name__ == "__main__":
  Test("Test1", "100", True)
  Test("Test2", "123.45e+6", True)
  Test("Test3", "+500", True)
  Test("Test4", "5e2", True)
  Test("Test5", "3.1416", True)
  Test("Test6", "600.", True)
  Test("Test7", "-.123", True)
  Test("Test8", "-1E-16", True)
  Test("Test9", "1.79769313486232E+308", True)

  Test("Test10", "12e", False)
  Test("Test11", "1a3.14", False)
  Test("Test12", "1+23", False)
  Test("Test13", "1.2.3", False)
  Test("Test14", "+-5", False)
  Test("Test15", "12e+5.4", False)
  Test("Test16", ".", False)
  Test("Test17", ".e1", False)
  Test("Test18", "e1", False)
  Test("Test19", "+.", False)
  Test("Test20", "", False)
  Test("Test21", None, False)

  Test("Test22", "2.e10", True)

File: 20_number_string/number_string.cpp
/*
 * 面试题20：表示数值的字符串
 * 题目：请实现一个函数用来判断字符串是否表示数值（包括整数和小数）。例如，
 * 字符串“+100”、“5e2”、“-123”、“3.1416”及“-1E-16”都表示数值，但“12e”、
 * “1a3.14”、“1.2.3”、“+-5”及“12e+5.4”都不是
 */
#include <iostream>
#include <cstdio>
using namespace std;

bool scanUnsignedInteger(const char** str);
bool scanInteger(const char** str);

/* Solution 数字形式A.BeC 整数部分，小数部分，以及指数部分 */
bool is_number(const char* str){
  if(str == nullptr)
    return false;

  // 判断整数部分
  bool numeric = scanInteger(&str);
  
  // 判断小数部分
  if(*str == '.'){
    ++str;

    // 下面一行代码用||的原因：
    // 1. 小数可以没有整数部分，例如.123等于0.123；
    // 2. 小数点后面可以没有数字，例如233.等于233.0；
    // 3. 当然小数点前面和后面可以有数字，例如233.666
    numeric = scanUnsignedInteger(&str) || numeric;
  }

  // 判断指数部分
  if(*str == 'E' || *str == 'e'){
    ++str;

    // 下面一行代码用&&的原因：
    // 1. 当e或E前面没有数字时，整个字符串不能表示数字，例如.e1、e1；
    // 2. 当e或E后面没有整数时，整个字符串不能表示数字，例如12e、12e+5.4
    numeric = numeric && scanInteger(&str);
  }

  return numeric && *str == '\0';
  
}

/* 判断是无符号整数 */
bool scanUnsignedInteger(const char** str){
  const char* before = *str;
  while(**str != '\0' && **str >= '0' && **str <= '9')
    ++(*str);

  return *str > before;
}

/* 判断是有符号整数 */
bool scanInteger(const char** str){
  const char* before = *str;
  if(**str == '+' || **str == '-')
    ++(*str);
  return scanUnsignedInteger(str);
}

/* 
 * 测试代码 
 */
void Test(const char* testName, const char* str, bool expected){
  if(testName != nullptr)
    printf("%s begins: ", testName);

  if(is_number(str) == expected)
    printf("passed.\n");
  else
    printf("failed.\n");
}

int main(int argc, char* argv[])
{
  Test("Test1", "100", true);
  Test("Test2", "123.45e+6", true);
  Test("Test3", "+500", true);
  Test("Test4", "5e2", true);
  Test("Test5", "3.1416", true);
  Test("Test6", "600.", true);
  Test("Test7", "-.123", true);
  Test("Test8", "-1E-16", true);
  Test("Test9", "1.79769313486232E+308", true);

  printf("\n\n");

  Test("Test10", "12e", false);
  Test("Test11", "1a3.14", false);
  Test("Test12", "1+23", false);
  Test("Test13", "1.2.3", false);
  Test("Test14", "+-5", false);
  Test("Test15", "12e+5.4", false);
  Test("Test16", ".", false);
  Test("Test17", ".e1", false);
  Test("Test18", "e1", false);
  Test("Test19", "+.", false);
  Test("Test20", "", false);
  Test("Test21", nullptr, false);

  Test("Test22", "2.e10", true);
  return 0;
}